### PR TITLE
Slice op

### PR DESCRIPTION
**Pull request by haijieg**
## Add slice opteration to sframe query planner
The slice operation can be directly applied to linear query plan,
or it will materialize (modifies) the existing query plan first.

Also add some utility function:
- is_linear_graph checks whether a query plan is linear
- make_sliced_graph slices the source node of a linear plan

## Make unity_sframe/unity_sarray::copy_range() more lazy
When step is 1, copy_range() will only sliced the query plan
instead of materializing. This affects all operations that
use copy_range(), e.g. and tail(), operator[], and make them much
faster.

- Fix the planner slice function to drop memoized length info.
- Change sframe callback implemenetation to use slice oeprator.

### CODE REVIEW

File: oss_src/sframe_query_engine/planning/planner.cpp [hunk 1]
```diff
@@ -52,7 +51,7 @@ static sframe execute_node_impl(pnode_ptr input_n, const materialize_options& ex
  * Executes a query plan, potentially parallelizing it if possible.
  * Also implements fast paths in the event the input node is a source node.
  */
-static sframe execute_node(pnode_ptr input_n, const materialize_options& exec_params) {
+static sframe execute_node(pnode_ptr input_n, materialize_options exec_params) {
```
Comment (ylow) [inline oss_src/sframe_query_engine/planning/planner.cpp]: why the change to copy?

File: oss_src/unity/extensions/additional_sframe_utilities.cpp [hunk 2]
```diff
@@ -18,28 +19,59 @@ typedef int(*sframe_callback_type)(const graphlab::flexible_type*, size_t, void*
...
+  std::vector<std::vector<std::vector<flexible_type>>> rows_by_segment_id;
+  rows_by_segment_id.resize(nthreads);
+  auto copy_out_callback = [&](size_t segment_id,
+                               const std::shared_ptr<sframe_rows>& data) {
+    for (auto& row: (*data)) {
+      rows_by_segment_id[segment_id].push_back(row);
+    }
+    return false;
+  };
+  sliced_input.materialize_to_callback(copy_out_callback, nthreads);
```
Comment (ylow) [inline oss_src/unity/extensions/additional_sframe_utilities.cpp]: this pulls the entire range into memory before triggering the callback. That might not be a good idea.

File: oss_src/unity/extensions/additional_sframe_utilities.cpp [hunk 3]
```diff
@@ -18,28 +19,59 @@ typedef int(*sframe_callback_type)(const graphlab::flexible_type*, size_t, void*
...
```
Comment (ylow) [inline oss_src/unity/extensions/additional_sframe_utilities.cpp]: Same as above

File: oss_src/sframe_query_engine/planning/planner.cpp (hunk 1 above)
Comment (haijieg) [inline oss_src/sframe_query_engine/planning/planner.cpp]: good catch. It was modified when I was trying to change the execution component and I forget to change it back. Will fix.

File: oss_src/unity/extensions/additional_sframe_utilities.cpp (hunk 2 above)
Comment (haijieg) [inline oss_src/unity/extensions/additional_sframe_utilities.cpp]: Agreed. However, the benefit is that the materialization is now done in parallel with out requiring the callback function to be reentrent. Since the callback is ranged, the risk of out-of-memory is limited and controllable.
